url888b6fd9b4_jmacglashan_commandsToTasks_tgz-pJ8-commands_tests_Model3TestsJ8/src/domain/singleagent/sokoban/SokobanDomain.java

```diff
@@ -0,0 +1,1345 @@
+				door.setValue(BOTTOMXATTNAME, Integer.parseInt(splitobject[3]));
+				door.setValue(BOTTOMYATTNAME, Integer.parseInt(splitobject[4]));
+				st.addObject(door);
+				doors++;
+			}
+			else if (splitobject[0].equals("block")){
+				ObjectInstance block = new ObjectInstance(SOKOBANDOMAIN.getObjectClass(BLOCKCLASS), BLOCKCLASS + blocks);
+				block.setValue(COLORATTNAME, colors.get(Integer.parseInt(splitobject[1])));
+				block.setValue(SHAPEATTNAME, shapes.get(Integer.parseInt(splitobject[2])));
+				block.setValue(XATTNAME, Integer.parseInt(splitobject[3]));
+				block.setValue(YATTNAME, Integer.parseInt(splitobject[4]));
+				st.addObject(block);
+				blocks++;
+			}
+			else if (splitobject[0].equals("agent")){
+				ObjectInstance agent = new ObjectInstance(SOKOBANDOMAIN.getObjectClass(AGENTCLASS), AGENTCLASS + agents);
+				agent.setValue(XATTNAME, Integer.parseInt(splitobject[1].trim()));
+				agent.setValue(YATTNAME, Integer.parseInt(splitobject[2].trim()));
+				st.addObject(agent);
+				agents++;
+			}
+		}
+		createMap(st);
+		return st;
+	}
+	public static void runTrajectory(SokobanDomain constructor, Domain domain, State st, String file){
+		BufferedReader in = null;
+		String currentState = null;
+		String trajectoryString = null;
+		String command = null;
+		Trajectory trajectory = null;
+		ArrayList<State> states = new ArrayList<State>();
+
+		try {
+			in = new BufferedReader(new FileReader(file));
+			command = in.readLine();
+			trajectoryString = in.readLine();
+			currentState = in.readLine();
+			while(currentState != null){
+				states.add(constructor.stringToState(currentState));
+				currentState = in.readLine();
+			}
+			in.close();
+		} catch (FileNotFoundException e) {
+			System.out.println("Error: File \"" + file + "\" not found.");
+			System.exit(1);
+		} catch (IOException e) {
+			e.printStackTrace();
+			System.exit(1);
+		}
+		ArrayList<GroundedAction> actions = new ArrayList<GroundedAction>();
+
+		ArrayList<String> stringActions = new ArrayList<String>(Arrays.asList(trajectoryString.split(" ")));
+
+		//perform actions from file
+		for (int i = 0; i < stringActions.size(); i++){
+			Action action = domain.getAction(stringActions.get(i));
+			actions.add(new GroundedAction(action, ""));
+			st = action.performAction(st, "");
+		}
+		trajectory = new Trajectory(states, actions);
+		TrainingElement element = new TrainingElement(command, trajectory);
+		//System.out.println(constructor.stateToString(element.trajectory.getState(element.trajectory.numStates() - 1)));
+		System.out.println("Trajectory completed");
+	}
+
+	public static int getMaxX(State st){
+		List<ObjectInstance> rooms = st.getObjectsOfTrueClass(ROOMCLASS);
+		int x = 0;
+		for (int i = 0; i < rooms.size(); i++){
+			ObjectInstance room = rooms.get(i); 
+			if (room.getDiscValForAttribute(BOTTOMXATTNAME) > x){
+				x = room.getDiscValForAttribute(BOTTOMXATTNAME);
+			}
+		}
+		return x+1;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1190:1198, 1202:1210.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.